newapi/ooniapi/auth.py

```diff
@@ -0,0 +1,213 @@
+    _send_email(dest_addr, msg)
+
+
+@metrics.timer("register_user")
+@auth_blueprint.route("/api/v1/register_user", methods=["POST"])
+def register_user():
+    """Auth Services: start email-based user registration
+    ---
+    parameters:
+      - in: body
+        name: register data
+        description: Registration data as HTML form or JSON
+        required: true
+        schema:
+          type: object
+          properties:
+            nickname:
+              type: string
+            email_address:
+              type: string
+    responses:
+      200:
+        description: Confirmation
+    """
+    log = current_app.logger
+    req = request.json if request.is_json else request.form
+    nick = req.get("nickname").strip()
+    # Accept all alphanum including unicode and whitespaces
+    if not nick.replace(" ", "").isalnum():
+        return jerror("Invalid user name")
+    if len(nick) < 3:
+        return jerror("User name is too short")
+    if len(nick) > 50:
+        return jerror("User name is too long")
+
+    email_address = req.get("email_address").strip().lower()
+    if not nick or not email_address:
+        return jerror("Invalid request")
+    if EMAIL_RE.fullmatch(email_address) is None:
+        return jerror("Invalid email address")
+
+    account_id = hash_email_address(email_address)
+    now = datetime.utcnow()
+    expiration = now + timedelta(days=1)
+    # On the backend side the registration is stateless
+    payload = {
+        "nbf": now,
+        "exp": expiration,
+        "aud": "register",
+        "account_id": account_id,
+        "email": email_address,
+        "nick": nick,
+    }
+    registration_token = create_jwt(payload)
+    log.info("sending registration token")
+    try:
+        send_login_email(email_address, nick, registration_token)
+        log.info("email sent")
+    except Exception as e:
+        log.error(e, exc_info=1)
+        return jerror("Unable to send the email")
+
+    return make_response(jsonify(msg="ok"), 200)
+
+
+@metrics.timer("user_login")
+@auth_blueprint.route("/api/v1/user_login", methods=["GET"])
+def user_login():
+    """Probe Services: login using a registration/login link
+    ---
+    parameters:
+      - name: k
+        in: query
+        type: string
+        description: JWT token with aud=register
+    responses:
+      200:
+        description: Login response, set cookie
+    """
+    log = current_app.logger
+    token = request.args.get("k")
+    try:
+        dec = decode_jwt(token, audience="register")
+        if dec.get("aud") != "register":
+            return jerror("Invalid token type", code=401)
+    except jwt.exceptions.InvalidSignatureError:
+        return jerror("Invalid credential signature", code=401)
+    except jwt.exceptions.DecodeError:
+        return jerror("Invalid credentials", code=401)
+
+    log.info("user login successful")
+    now = datetime.utcnow()
+    payload = {
+        "nbf": now,
+        "iat": now,
```

Comment:
As discussed we should be adding some sort of expiration time to these tokens to not make them live forever.

Reply:
Done